src/puppetlabs/puppetdb/command.clj

```diff
@@ -302,11 +303,45 @@
 (def supported-command?
   (comp (kitchensink/valset command-names) :command))
 
+(defprotocol PuppetDBCommandDispatcher
+  (enqueue-command [this connection endpoint command version payload])
+  (enqueue-raw-command [this connection endpoint raw-command])
+  (stats [this]))
+
 (defservice command-service
+  PuppetDBCommandDispatcher
   [[:PuppetDBServer shared-globals]
    [:MessageListenerService register-listener]]
+  (init [this context]
+    (assoc context :stats (atom {:received-commands 0
+                                 :executed-commands 0})))
   (start [this context]
-         (let [{:keys [scf-write-db catalog-hash-debug-dir]} (shared-globals)]
-           (register-listener supported-command? #(process-command! % {:db scf-write-db
-                                                                       :catalog-hash-debug-dir catalog-hash-debug-dir}))
-           context)))
+    (let [{:keys [scf-write-db catalog-hash-debug-dir]} (shared-globals)
+          config {:db scf-write-db
+                  :catalog-hash-debug-dir catalog-hash-debug-dir}]
+      (register-listener
+       supported-command?
+       (fn [cmd]
+         (let [result (process-command! cmd config)]
+           (swap! (:stats context)
+                  (fn [cur] (update cur :executed-commands inc)))
+           result)))
+      context))
+
+  (stats [this]
+    @(:stats (service-context this)))
+
+  (enqueue-command [this connection endpoint command version payload]
+    (let [result (do-enqueue-command connection endpoint
```

Comment:
Do we want foo or foo! -- note that we have submit-command, not submit-command! in http.commands, so wasn't sure.